### PR TITLE
[Inductor] Fix argmax/argmin returning incorrect indices for boolean tensors on CUDA

### LINKED ISSUE #174069: [Inductor] argmax/max returns incorrect indices for boolean tensors on CUDA
## Describe the bug

When using `torch.compile` with the Inductor backend on CUDA, `max(dim)[1]` (argmax) returns incorrect indices for boolean tensors. The eager backend returns correct results.

## Minimal Reproduction

```python
import torch
import torch._dynamo

device = 'cuda'

def fn(x):
    return x.max(1, keepdim=True)[1]

# Boolean tensor with True starting at position 4
mask = torch.zeros(4, 8, 32, device=device, dtype=torch.bool)
mask[:, 4:, :] = True

result_eager = fn(mask)
torch._dynamo.reset()
result_compiled = torch.compile(fn, backend='inductor')(mask)

print(f'Expected index: 4')
print(f'Eager result: {result_eager[0, 0, 0].item()}')
print(f'Inductor result: {result_compiled[0, 0, 0].item()}')
# Output:
# Expected index: 4
# Eager result: 4
# Inductor result: 1  # WRONG!
```

## Results

| Tensor dtype | Eager | Inductor | Correct? |
|--------------|-------|----------|----------|
| bool | 4 | 1 | No |
| float | 4 | 4 | Yes |
| int | 4 | 4 | Yes |

The bug only affects boolean tensors.

## Impact

This bug affects any code that uses `argmax` or `max(...)[1]` on boolean tensors when compiled with Inductor on CUDA. For example, it can cause incorrect behavior in `cumprod_backward` when using composite-compliant implementations with `at::where`.

## Versions

- PyTorch: 2.11.0a0 (main branch, commit a3bda4952b3)
- CUDA: 12.5

cc @ezyang @gchanan @kadeng @msaroufim @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @muchulee8 @amjames @aakhundov @coconutruben @jataylo @isuruf

### PR DESCRIPTION
**Pull request by Aminsed**
## Summary

When using `torch.compile` with the Inductor backend on CUDA, `max(dim)[1]` (argmax) and `min(dim)[1]` (argmin) returned incorrect indices for boolean tensors. This was because Triton's comparison operators don't work correctly on boolean types.

The fix casts boolean tensors to int32 before performing argmax/argmin reductions in the lowering pass.

## Test Plan

Tested locally on CUDA:

```python
import torch
import torch._dynamo

def fn(x):
    return x.max(1, keepdim=True)[1]

mask = torch.zeros(4, 8, 32, device='cuda', dtype=torch.bool)
mask[:, 4:, :] = True

# Before fix: Inductor returned 1, should be 4
# After fix: Inductor correctly returns 4
result_eager = fn(mask)
result_compiled = torch.compile(fn, backend='inductor')(mask)
assert torch.equal(result_eager, result_compiled)
```

Fixes https://github.com/pytorch/pytorch/issues/174069

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @mlazos @ezyang @jansel @Chillee @bdhirsh

### CODE REVIEW

**Comment by ngimel** [comment]
Please add a test

**Review by ngimel — APPROVED** [review]

**Review by jansel — APPROVED** [review]

**Review by jansel — CHANGES_REQUESTED** [review]
The `test_max_min_bool_dynamic_shapes_cpu` test failures looks related.

**Review by mlazos — APPROVED** [review]

**Review by jansel — APPROVED** [review]

**Comment by jansel** [comment]
Is that test failure real?
